parser/yaml_output_functions.c

```diff
@@ -0,0 +1,313 @@
+ * \param l3vals The values corresponding to l3keys
+ */
+void write_yaml_from_struct_3 (char *yamlname, int asize, struct fmsyamloutkeys *topkeys, struct fmsyamloutvalues *topvals, int a2size, struct fmsyamloutkeys *l2keys, struct fmsyamloutvalues *l2vals, int a3size, int * n3each, struct fmsyamloutkeys *l3keys, struct fmsyamloutvalues *l3vals){
+  yaml_emitter_t emitter; /* libyaml emitter */
+  yaml_event_t event; /* libyaml event for the output yaml */
+  int s3count = 0; /* A counter to keep track of the number of level 3 arrays output */
+  FILE * yamlout; /* The file for the YAML output. */
+  /* open the yaml output file. Only 1 core should do this */
+    yamlout = fopen(yamlname,"w");
+  /* Start the emmitter */  
+    yaml_emitter_initialize(&emitter);
+    yaml_emitter_set_output_file(&emitter, yamlout);
+
+    yaml_stream_start_event_initialize(&event, YAML_UTF8_ENCODING);
+    if (!yaml_emitter_emit(&emitter, &event)) goto error;
+
+    yaml_document_start_event_initialize(&event, NULL, NULL, NULL, 0);
+    if (!yaml_emitter_emit(&emitter, &event)) goto error;
+/* start the event (top level) */
+    yaml_mapping_start_event_initialize(&event, NULL, (yaml_char_t *)YAML_MAP_TAG,
+        1, YAML_ANY_MAPPING_STYLE);
+    if (!yaml_emitter_emit(&emitter, &event)) goto error;
+
+/* write the top level */
+ write_keys_vals_yaml (&emitter, &event , 0, topkeys, topvals);
+
+ /* Check for the next level key */
+ if (topkeys->level2key[0] !='\0') {
+  /* Start the secodn level event */
+  yaml_scalar_event_initialize(&event, NULL, (yaml_char_t *)YAML_STR_TAG,
+      (yaml_char_t *)topkeys->level2key, strlen(topkeys->level2key), 1, 0, 
+      YAML_PLAIN_SCALAR_STYLE);
+  if (!yaml_emitter_emit(&emitter, &event)) goto error;
+  /* Start the sequencing */
+  yaml_sequence_start_event_initialize(&event, NULL, (yaml_char_t *)YAML_SEQ_TAG,
+      1, YAML_ANY_SEQUENCE_STYLE);
+  if (!yaml_emitter_emit(&emitter, &event)) goto error;
+  /* loop through the structs */
+  for (int s2 = 0 ; s2 < a2size ; s2++){
+    	yaml_mapping_start_event_initialize(&event, NULL, (yaml_char_t *)YAML_MAP_TAG,
+        	1, YAML_ANY_MAPPING_STYLE);
+    	if (!yaml_emitter_emit(&emitter, &event)) goto error;
+    	/* call the write function */
+    	write_keys_vals_yaml (&emitter, &event , s2, l2keys, l2vals);
+    	/* Next level keys */
+    	if (l2keys->level2key[0] !='\0') {
+     		/* Start the third level event */
+     		yaml_scalar_event_initialize(&event, NULL, (yaml_char_t *)YAML_STR_TAG,
+     		(yaml_char_t *)l2keys->level2key, strlen(l2keys->level2key), 1, 0,
+     		YAML_PLAIN_SCALAR_STYLE);
+     		if (!yaml_emitter_emit(&emitter, &event)) goto error;
+     		/* Start the sequencing */
+     		yaml_sequence_start_event_initialize(&event, NULL, (yaml_char_t *)YAML_SEQ_TAG,
+         		1, YAML_ANY_SEQUENCE_STYLE);
+     		if (!yaml_emitter_emit(&emitter, &event)) goto error;
+    		/* loop through the structs */
+		int s3start = s3count;
```

Comment:
needs indents (I originally thought this variable declaration was in the wrong spot... too much fortran)